File: backend/aws-api/user-crud/PutController.py
import json
import hashlib
import rds_config

def put(connection, data):
    cursor = connection.cursor()
    body = json.loads(data)
    
    print("=====testing=====")
    print(body)
    
    username = body["userName"]
    email = body["email"]
    membership= body["membership"]
    password = body["password"]
    key = rds_config.user_password_key
    iv = body["iv"]
    print(password)
    print(key)
    print(iv)
    
    # m = hashlib.sha256()
    # m.update(password.encode('utf-8'))
    # m.digest()
    # print(m)
    
    # cipher = AES.new(key, AES.MODE_CBC, iv=iv)
    # originalData = unpad(cipher.decrypt(cipheredData), AES.block_size)
    # print(originalData)
    
    # sql = "INSERT INTO users "
    # sql += "(tree_name_en, tree_name_cn, tree_alias, scientific_name, family_code, ecologic, cap_96, cap_586, hk_rare, cn_rare, flowering, fruit, tree_desc_en, tree_desc_cn, tree_image) "
    # sql += "VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)"
    # val = (
    #     body["treeNameEn"],
    #     body["treeNameCn"],
    #     body["treeAlias"],
    #     body["scientificName"],
    #     body["familyCode"],
    #     body["ecologic"],
        
    #     body["cap96"],
    #     body["cap586"],
    #     body["hkRare"],
    #     body["cnRare"],
        
    #     body["flowering"],
    #     body["fruit"],
        
    #     body["treeDescEn"],
    #     body["treeDescCn"],
    #     body["treeImage"]
        
    # )
    
    statusCode = 502
    message = "Create user record failure"
    # try:
    #     cursor.execute(sql, val)
    #     connection.commit()
    #     statusCode = 200
    #     message = "Create tree record succeeded "
    # except:
    #     connection.rollback()
    
    return {
        'statusCode': 200,
        'headers': {
            'Content-Type': 'application/json',
            'Access-Control-Allow-Headers': 'Content-Type',
            'Access-Control-Allow-Origin': '*',
            'Access-Control-Allow-Methods': 'OPTIONS,PUT'
        },
        'body': message
    }
    

File: backend/aws-api/user-crud/PutController.js
const { decrypt, hash } = require("CryptoUtil");

exports.put = function(pool, body, callback) {
    let obj = JSON.parse(body);
    switch(obj["requestType"]) {
        case "updateDetail":
            updateDetail(pool, obj, callback);
            break;
        case "changePassword":
            changePassword(pool, obj, callback);
            break;
        case "login":
            login(pool, obj, callback);
            break;
    }
};

function updateDetail(pool, data, callback) {
    let userId = data["userId"],
        userName = data["userName"],
        email = data["email"],
        membership = data["membership"],
        sql = `UPDATE users SET user_name='${userName}', email='${email}', membership=${membership} WHERE email=${email}`;
        
    pool.getConnection(function(err, connection) {
        if(err) throw err;
        connection.query(sql, function(error, results) {
            if(error) {
                callback(error);
            } else {
                const result = {
                    'statusCode': 200,
                    'headers': {
                        'Content-Type': 'application/json',
                        'Access-Control-Allow-Headers': 'Content-Type',
                        'Access-Control-Allow-Origin': '*',
                        'Access-Control-Allow-Methods': 'OPTIONS,PUT'
                    },
                    'body': "Update details success"
                };
                callback(null, result);
            }
        });
    });
}

function changePassword(pool, data, callback) {
    let email = data["email"],
        key = process.env.PASSWORD_KEY;
        
    let oldPassword = data["oldPassword"],
        oldPasswordIV = data["oldPasswordIV"],
        decryptedOldPassword = decrypt(oldPassword, key, oldPasswordIV),
        hashedOldPassword = hash(decryptedOldPassword);
        
    let newPassword = data["newPassword"],
        newPasswordIV = data["newPasswordIV"],
        decryptedNewPassword = decrypt(newPassword, key, newPasswordIV),
        hashedNewPassword = hash(decryptedNewPassword);
        
    let checkingSql = `SELECT * FROM users WHERE email=${email}`,
        updateSql = `UPDATE users SET password='${hashedNewPassword}' WHERE email=${email}`;
    
    pool.getConnection(function(err, connection) {
        if(err) throw err;
        
        connection.query(checkingSql, function(error, results) {
            if(error) {
                callback(error);
            } else {
                let recordPassword = results[0]["password"];
                if(recordPassword === hashedOldPassword) {
                    connection.query(updateSql, function(error2, results2) {
                        if(error2) {
                            callback(error2);
                        } else {
                            const result = {
                                'statusCode': 200,
                                'headers': {
                                    'Content-Type': 'application/json',
                                    'Access-Control-Allow-Headers': 'Content-Type',
                                    'Access-Control-Allow-Origin': '*',
                                    'Access-Control-Allow-Methods': 'OPTIONS,PUT'
                                },
                                'body': "Password changed success"
                            };
  
                            callback(null, result);
                        }
                    });
                } else {
                    const result = {
                        'statusCode': 401,
                        'headers': {
                            'Content-Type': 'application/json',
                            'Access-Control-Allow-Headers': 'Content-Type',
                            'Access-Control-Allow-Origin': '*',
                            'Access-Control-Allow-Methods': 'OPTIONS,PUT'
                        },
                        'body': "Old password Unauthorized"
                    };
                    callback(null, result);
                }
            }
        });
    });
}

function login(pool, data, callback) {
    let email = data["email"],
        key = process.env.PASSWORD_KEY,
        password = data["password"],
        iv = data["iv"],
        decrypted = decrypt(password, key, iv),
        hashed = hash(decrypted),
        checkingSQL = `SELECT * FROM users WHERE email='${email}'`;
        
    pool.getConnection(function(err, connection) {
        if(err) throw err;
        
        connection.query(checkingSQL, function(error, results) {
            if(error) {
                callback(error);
            } else {
                console.log("results ", results);
                let userName = results[0]["user_name"],
                    membership = results[0]["membership"];
                    
                let recordPassword = results[0]["password"];
                if(recordPassword === hashed) {
                    let timestamp = Date.now(),
                        setTimeSQL = `UPDATE users SET login='${timestamp}' WHERE email='${email}'`;
                    
                    connection.query(setTimeSQL, function(error2, results2) {
                        console.log("error", error2);
                        if(error2) {
                            callback(error2);
                        } else {
                            const body = JSON.stringify({
                                message: "Login success",
                                userName: userName,
                                email: email,
                                membership: membership,
                                timestamp: timestamp
                            });
                            const result = {
                                'statusCode': 200,
                                'headers': {
                                    'Content-Type': 'application/json',
                                    'Access-Control-Allow-Headers': 'Content-Type',
                                    'Access-Control-Allow-Origin': '*',
                                    'Access-Control-Allow-Methods': 'OPTIONS,PUT'
                                },
                                'body': body
                            };
                            console.log("xxxxxx result", result);
                            callback(null, result);
                        }
                    });
                } else {
                    const result = {
                        'statusCode': 401,
                        'headers': {
                            'Content-Type': 'application/json',
                            'Access-Control-Allow-Headers': 'Content-Type',
                            'Access-Control-Allow-Origin': '*',
                            'Access-Control-Allow-Methods': 'OPTIONS,PUT'
                        },
                        'body': "User Unauthorized"
                    };
                    callback(null, result);
                }
            }
        });
    });
}